### PR TITLE
[WIP] ST examples and related code edit

### PR DESCRIPTION
**Pull request by LittleChenCc**
### PR types
New feature
### PR changes
Others
### Describe
ST examples and related code edit

### CODE REVIEW

File: deepspeech/exps/u2/model.py [hunk 1]
```diff
@@ -414,6 +418,53 @@ def compute_metrics(self, audio, audio_len, texts, texts_len, fout=None):
...
+
+        for target, result in zip(refs, hyps):
+            len_refs += len(target.split())
+            num_ins += 1
+            if fout:
+                fout.write(result + "\n")
+            logger.info("\nReference: %s\nHypothesis: %s" %
+                        (target, result))
+            logger.info("One example BLEU = %s" %
+                        (bleu_func([result], [[target]]).prec_str))
```
Comment (zh794390558) [inline deepspeech/exps/u2/model.py]: 可以抽象下，不直接暴露bleu_func = sacrebleu.corpus_bleu

File: deepspeech/exps/u2/model.py [hunk 2]
```diff
@@ -21,13 +21,14 @@
 from typing import Optional
 
 import numpy as np
+import sacrebleu
```
Comment (zh794390558) [inline deepspeech/exps/u2/model.py]: 可以单独copy一份u2到st，代码可以copy，隔离重要

File: deepspeech/exps/u2/model.py [hunk 3]
```diff
@@ -21,13 +21,14 @@
...
 import paddle
 from paddle import distributed as dist
 from paddle.io import DataLoader
 from yacs.config import CfgNode
 
 from deepspeech.io.collator import SpeechCollator
-from deepspeech.io.dataset import ManifestDataset
+from deepspeech.io.dataset import ManifestDataset, FeaturizedManifestDataset
```
Comment (zh794390558) [inline deepspeech/exps/u2/model.py]: KaldiManifestDataset 更清晰。

File: deepspeech/exps/u2/model.py [hunk 4]
```diff
@@ -262,14 +266,14 @@ def setup_dataloader(self):
...
-        test_dataset = ManifestDataset.from_config(config)
+        test_dataset = Dataset.from_config(config)
         # return text ord id
         self.test_loader = DataLoader(
             test_dataset,
             batch_size=config.decoding.batch_size,
             shuffle=False,
             drop_last=False,
-            collate_fn=SpeechCollator(keep_transcription_text=True))
+            collate_fn=SpeechCollator(keep_transcription_text=False))
```
Comment (zh794390558) [inline deepspeech/exps/u2/model.py]: keep_transcription_text必须是True

File: deepspeech/exps/u2/model.py (hunk 3 above)
Comment (zh794390558) [reply deepspeech/exps/u2/model.py]: #657 #665 可以关注下，在修改 datapipe

File: deepspeech/exps/u2/model.py [hunk 5]
```diff
@@ -414,6 +418,53 @@ def compute_metrics(self, audio, audio_len, texts, texts_len, fout=None):
             num_frames=audio_len.sum().numpy().item(),
             decode_time=decode_time)
 
+    def compute_translation_metrics(self, audio, audio_len, texts, texts_len, fout=None):
```
Comment (zh794390558) [inline deepspeech/exps/u2/model.py]: 这部分应该是可以和 compute metrics 合并的，只是多计算个 metric 而已。

File: deepspeech/exps/u2/model.py (hunk 1 above)
Comment (zh794390558) [reply deepspeech/exps/u2/model.py]: 期望和wer cer 接口兼容。这样可以同时计算多个metric，需要修改下保存metric中间统计量的数据结构。

File: deepspeech/exps/u2/model.py [hunk 6]
```diff
@@ -422,59 +473,110 @@ def test(self):
...
-                num_ins,
-                "err_sum":
-                errors_sum,
-                "ref_len":
-                len_refs,
-                "decode_method":
-                self.config.decoding.decoding_method,
-            })
-            f.write(data + '\n')
+        if self.config.decoding.error_rate_type == 'bleu':
```
Comment (zh794390558) [inline deepspeech/exps/u2/model.py]: 不需要区分bleu，流程是一样的。

File: deepspeech/frontend/utility.py [hunk 7]
```diff
@@ -42,7 +42,8 @@ def read_manifest(
         max_output_len=float('inf'),
         min_output_len=0.0,
         max_output_input_ratio=float('inf'),
-        min_output_input_ratio=0.0, ):
+        min_output_input_ratio=0.0, 
+        keep_all=False):
```
Comment (zh794390558) [inline deepspeech/frontend/utility.py]: 默认的参数就是keep all的

File: deepspeech/io/dataset.py [hunk 8]
```diff
@@ -30,13 +31,16 @@
...
+    "FeaturizedManifestDataset"
 ]
 
 logger = Log(__name__).getlog()
 
 # namedtupe need global for pickle.
 TarLocalData = namedtuple('TarLocalData', ['tar2info', 'tar2object'])
 
+# filter ignore
+IgnoreSet = ['dev', 'test']
```
Comment (zh794390558) [inline deepspeech/io/dataset.py]: 这部分不需要，还是train test 匹配最好。放开的话可能用户感知不到。会出bug。

File: deepspeech/io/dataset.py [hunk 9]
```diff
@@ -348,3 +354,27 @@ def __len__(self):
     def __getitem__(self, idx):
         instance = self._manifest[idx]
         return self.process_utterance(instance["feat"], instance["text"])
+
+class FeaturizedManifestDataset(ManifestDataset):
```
Comment (zh794390558) [inline deepspeech/io/dataset.py]: 现在可以这样，后面需要关注下 #657 #665 这两个PR。我理解可以不用加 FeaturizedManifestDataset 这个了。

File: examples/mustc/conf/augmentation.json [hunk 10]
```diff
@@ -0,0 +1,34 @@
+[
+  {
+    "type": "shift",
+    "params": {
+      "min_shift_ms": -5,
+      "max_shift_ms": 5
+    },
+    "prob": 1.0
+  },
+  {
```
Comment (zh794390558) [inline examples/mustc/conf/augmentation.json]: 先speed后shift。

File: examples/mustc/conf/transformer.yaml [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+  featurized: True
+  sortagrad: True 
+  shuffle_method: batch_shuffle
+  num_workers: 2
+
+
+# network architecture
+model:
+    cmvn_file: None
+    # cmvn_file_type: "json"
```
Comment (zh794390558) [inline examples/mustc/conf/transformer.yaml]: 看下正确的配置吧。

File: examples/mustc/conf/transformer.yaml [hunk 12]
```diff
@@ -0,0 +1,107 @@
+# https://yaml.org/type/float.html
+data:
+  train_manifest: data/manifest.train
+  dev_manifest: data/manifest.dev
+  test_manifest: data/manifest.test
+  vocab_filepath: data/vocab.txt
+  unit_type: 'spm'
+  spm_model_prefix: 'data/train_sp.en-de.de_bpe8000_tc'
+  mean_std_filepath: ""
+  # augmentation_config: conf/augmentation.json
```
Comment (zh794390558) [inline examples/mustc/conf/transformer.yaml]: 可以把aug文件的prob都改成0.

File: examples/mustc/path.sh [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+export LC_ALL=C
+
+# Use UTF-8 in Python to avoid UnicodeDecodeError when LC_ALL=C
+export PYTHONIOENCODING=UTF-8 
+export PYTHONPATH=${MAIN_ROOT}:${PYTHONPATH}
+
+export LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:/usr/local/lib/
+
+
+MODEL=u2
```
Comment (zh794390558) [inline examples/mustc/path.sh]: st 单独加个目录吧。

File: examples/mustc/run.sh [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+# hand-craft data process
+
+if [ ${stage} -le 1 ] && [ ${stop_stage} -ge 1 ]; then
+    # train model, all `ckpt` under `exp` dir
+    CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7,8,9 ./local/train.sh ${conf_path}  ${ckpt}
+fi
+
+if [ ${stage} -le 2 ] && [ ${stop_stage} -ge 2 ]; then
+    # avg n best model
+    ../../utils/avg.sh exp/${ckpt}/checkpoints ${avg_num}
```
Comment (zh794390558) [inline examples/mustc/run.sh]: 直接 avg.sh 可以调用。

**Review by zh794390558 — COMMENTED** [review]
代码修改后运行 pre-commit run -a

File: deepspeech/exps/u2/model.py (hunk 4 above)
Comment (LittleChenCc) [reply deepspeech/exps/u2/model.py]: yes, I understood. I changed this one for evaluating bleu since I read 'ids' from the preprocessed data. I will modify the code to be consistent with the current data pipeline.

File: deepspeech/exps/u2/model.py (hunk 3 above)
Comment (LittleChenCc) [reply deepspeech/exps/u2/model.py]: Got it. 我会关注下这两个PR.

File: deepspeech/exps/u2/model.py (hunk 5 above)
Comment (LittleChenCc) [reply deepspeech/exps/u2/model.py]: 我理解是否还是应该分离开，因为实际还是两种任务。比如WER相对于ST是无效指标，应该忽略不算，BLEU之于ASR也是一样。
